src/transformStateWithClones.js

```diff
@@ -7,7 +7,60 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateArr = [];
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+    const newState = stateArr.length
+      ? { ...stateArr[stateArr.length - 1] }
+      : { ...state };
+
+    switch (type) {
+      case 'addProperties':
+        stateArr.push(Object.assign(newState, extraData));
+        break;
+
+      case 'removeProperties':
+        if (!keysToRemove.length) {
+          stateArr.push(newState);
+          break;
+        }
+
+        for (const key of keysToRemove) {
+          if (Object.hasOwn(newState, key)) {
```

Comment:
The `if` condition `Object.hasOwn(newState, key)` is redundant since the `delete` operation will not throw an error even if the key does not exist on the object. You can safely remove this condition. [CODE STYLE]